### PR TITLE
Refactor/task domain events & commands

### PR DESCRIPTION
**Pull request by fredericvilcot**
This PR adds a massive refactor to the `task` domain, follows https://github.com/okp4/ui/pull/476.

The idea is to implement some principles inherited both from Event Sourcing and CQRS patterns.

In particular, it highlights the necessity to enable events snapshots to mutate the state and to trigger those events by commands that can be invoked by any controller.

When the command is invoked, this is its responsibility to ensure that payloads are well designed and uncorrupted, and this in order to preserve the unicity, immutability and infallibility of the generated event that will create or update the state.

Consequently, it allows to protect entities in the domain and prevent from leaking them through commands, events or selectors.

The next step will be to add a command bus that will execute asynchronous operations and will allow to decorrelate commands from redux.

### CODE REVIEW

**Review by amimart — APPROVED** [review]

File: src/domain/task/helper/test.helper.ts [hunk 1]
```diff
@@ -1,22 +1,55 @@
...
+          task: {
+            byId: foundTask ? acc.task.byId.remove(cur.id) : acc.task.byId,
+            byType: acc.task.byType
+              .map((value: Readonly<OrderedSet<string>>) => value.delete(cur.id))
+              .filter((value: Readonly<OrderedSet<string>>) => !value.isEmpty())
+          },
+          displayedTaskIds: acc.displayedTaskIds.remove(cur.id)
+        }
+      : acc
+  }, iniialState)
```
Comment (JoeyDjr) [inline src/domain/task/helper/test.helper.ts:43]:
```suggestion
tasks.reduce<AppState>((acc: DeepReadonly<AppState>, cur: Task, index: number): AppState => 
  index !== errorIndex
    ? {
        task: {
          byId: acc.task.byId.get(cur.id) ? acc.task.byId.remove(cur.id) : acc.task.byId,
          byType: acc.task.byType
            .map((value: Readonly<OrderedSet<string>>) => value.delete(cur.id))
            .filter((value: Readonly<OrderedSet<string>>) => !value.isEmpty())
        },
        displayedTaskIds: acc.displayedTaskIds.remove(cur.id)
      }
    : acc
, iniialState)
```

It's a suggestion It's up to you

Comment (fredericvilcot) [reply src/domain/task/helper/test.helper.ts:43]: @JoeyDjr 
Thanks for you suggestion to be more functional, if you don't mind I will keep the original version as the code is only executed in `test` files, so I think the modification is not relevant here.

Comment (JoeyDjr) [reply src/domain/task/helper/test.helper.ts:43]: ok for me

**Review by JoeyDjr — APPROVED** [review]

**Review by ErikssonJoakim — APPROVED** [review]
This is some impressive code! 👍

File: src/domain/task/helper/test.helper.ts [hunk 2]
```diff
@@ -1,22 +1,55 @@
...
+): EventParameter<T> => {
   return [
     { ...{ type }, ...{ payload } },
     { initiator, timestamp: date }
   ]
 }
+
+// Compute state after invoking removeTask* commands
+export const getExpectedStateAfterRemove = (
+  iniialState: DeepReadonly<AppState>,
```
Comment (lolottetheclash) [inline src/domain/task/helper/test.helper.ts:26]: Typo here ;)

File: src/domain/task/usecase/amend-task-status/amendTaskStatus.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+export const amendTaskStatus =
+  (amendTaskStatusPayload: DeepReadonly<AmendTaskStatus>): ThunkResult<Promise<void>> =>
+  // eslint-disable-next-line @typescript-eslint/typedef
+  async (dispatch, getState) => {
+    const { id, timestamp = new Date(), status }: AmendTaskStatus = amendTaskStatusPayload
+    if (!getState().task.byId.has(id)) {
+      dispatchError(
+        new UnspecifiedError(
+          `Oops.. The provided id '${id}' does not exist, so we can't amend the status of an unknwom task..`
```
Comment (lolottetheclash) [inline src/domain/task/usecase/amend-task-status/amendTaskStatus.ts:22]: typo here => `unknwom`

**Review by lolottetheclash — APPROVED** [review]
Such a great job !!! 💪

**Review by ccamel — APPROVED** [review]

File: src/domain/task/helper/test.helper.ts (hunk 2 above)
Comment (fredericvilcot) [reply src/domain/task/helper/test.helper.ts:26]: Good catch!

**Pull request merged**